### DISCUSSION TITLE
Egress using Kubernetes Gateway API

### DISCUSSION THREAD

**Discussion by CerqueiraYYZ** [istio/istio#44757] [Q&A] [answered]
Hello, I'm wondering if there is a way to migrate Egress Gateway implementation on Istio using Kubernetes Gateway API with Istio GateWayClass.

Is it possible to have a clear example?
Here is an example of my service entry Egress gateway on istio.

'apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: mock
  namespace: istio-system
spec:
  hosts:
  - dev-example-eks.alb.example.com
  ports:
  - number: 80
    name: http
    protocol: HTTP
  - number: 443
    name: https
    protocol: HTTPS
  resolution: DNS
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mock-egressgateway
  namespace: istio-system
spec:
  selector:
    app: mock-egressgateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - dev-example-eks.alb.example.com
    tls:
      mode: ISTIO_MUTUAL
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: egressgateway-for-mock-dev
  namespace: istio-system
spec:
  host: mock-egressgateway.istio-system.svc.cluster.local
  subsets:
  - name: mock
    trafficPolicy:
      loadBalancer:
        simple: ROUND_ROBIN
      portLevelSettings:
      - port:
          number: 443
        tls:
          mode: ISTIO_MUTUAL
          sni: dev-example-eks.alb.example.com
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: direct-mock-dev-through-egress-gateway
  namespace: istio-system
spec:
  hosts:
  - dev-example-eks.alb.example.com
  gateways:
  - mock-egressgateway
  - mesh
  http:
  - name: mock-route1-dev
    match:
    - gateways:
      - mesh
      port: 80
    route:
    - destination:
        host: mock-egressgateway.istio-system.svc.cluster.local
        subset: mock
        port:
          number: 443
      weight: 100
  - name: mock-route2-dev
    match:
    - uri:
        prefix: /mock/skybridge
    - gateways:
      - mock-egressgateway
      port: 443
    route:
    - destination:
        host: dev-example-eks.alb.example.com
        port:
          number: 443
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: originate-tls-for-mock-dev
  namespace: istio-system
spec:
  host: dev-example-eks.alb.example.com
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
    portLevelSettings:
    - port:
        number: 443
      tls:
        mode: SIMPLE
        sni: dev-example-eks.alb.example.com'

**Comment by kfaseela** [comment]
cc @frankbu

**Comment by frankbu** [comment, marked as answer]
Configuring egress traffic is not supported in Gateway API yet, but it is being discussed and will be added in a future GEP: https://github.com/kubernetes-sigs/gateway-api/issues/1856

In the meantime, the Istio recommendation is to use Istio configuration resources for things not currently supported in Gateway API. We plan to add some documentation soon to show how Istio and Gateway API can be used together. For example, to use Gateway API to define a gateway, but use an Istio VirtualService (instead of HTTPRoute) to define the routes in cases where some required feature of Istio is not currently supported in Gateway API. Egress is another example where Istio configuration resources should be used for now.

**Reply by CerqueiraYYZ** [reply]
Thanks a lot, Frank, I'll keep an eye on future GEP.

**Comment by gberche-orange** [comment]
@frankbu https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway/ mentions that the istio egress gateway relies on experimental gep(s), but does this list it/them

https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway/
> The following instructions allow you to choose to use either the Gateway API or the Istio configuration API when configuring traffic management in the mesh. Follow instructions under either the Gateway API or Istio APIs tab, according to your preference.
>This document configures Istio using Gateway API features that are experimental 

Can you please help point which of the current experimental GEP(s) is refered to among the current list at https://gateway-api.sigs.k8s.io/geps/list/ reproduced below ? 

I would assume it to be [GEP-4894: Backend Resource](https://gateway-api.sigs.k8s.io/geps/gep-4894/)

however https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway/ seems to use istio-specific CR

```yaml
apiVersion: networking.istio.io/v1
kind: ServiceEntry
```

instead of GEP experimental
```yaml
apiVersion: gateway.networking.k8s.io/v1alpha1
kind: Backend
```

**Comment by howardjohn** [comment]
I think that is TLSRoute. Which is no longer experimental; we should update
it to use 'v1'
